contracts/L1GlobalForkRequester.sol

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity ^0.8.20;
+
+/* 
+   Contract to proxy a fork request from the bridge to its ForkingManager
+   Any L2 contract can call us over the bridge to get us to request a fork.
+   We record the dispute they were forking over
+*/
+
+// TODO: An alternative would be to look up the ForkingManager on L2ChainInfo and call it directly.
+// TODO: We could use this to manage the dispute data, in which case we could gate initiateFork() to only be callable by us.
+// TODO: The whitepaper implies a whitelist about who is allowed to fork us. Currently anybody can as long as they pay.
+
+import {IForkingManager} from "./interfaces/IForkingManager.sol";
+import {IPolygonZkEVMBridge} from "@RealityETH/zkevm-contracts/contracts/interfaces/IPolygonZkEVMBridge.sol";
+import {IPolygonZkEVM} from "@RealityETH/zkevm-contracts/contracts/interfaces/IPolygonZkEVM.sol";
+
+// NB We'd normally use the interface IForkableBridge here but it causes an error:
+//  Error (5005): Linearization of inheritance graph impossible
+import {ForkableBridge} from "./ForkableBridge.sol";
+
+contract L1GlobalForkRequester {
+
+    // Any bridge (or any contract pretending to be a bridge) can call this.
+    // We'll look up its ForkingManager and ask it for a fork.
+    // TODO: It might make more sense if this contract requested the chain ID then kept a record of it.
+    // ...then the ForkingManager would be locked to only fork based on our request.
+
+    function onMessageReceived(address _originAddress, uint32 _originNetwork, bytes memory _data) external payable {
+
+      ForkableBridge bridge = ForkableBridge(msg.sender);
+      IForkingManager fm = IForkingManager(bridge.forkmanager());
+
+      // The chain ID should always be the chain ID expected by the ForkingManager.
+      // It shouldn't be possible for it to be anything else but we'll check it anyhow.
+      IPolygonZkEVM zkevm = IPolygonZkEVM(fm.zkEVM());
+      require(uint64(_originNetwork) == zkevm.chainID(), "Bad _originNetwork, WTF");
+    
+      // We also check in the opposite direction to make sure the ForkingManager thinks the bridge is its bridge
+      require(address(fm.bridge()) == msg.sender, "Bridge mismatch, WTF");
+
+      // TODO:
+      // 1) Work out what happens if this reverts.
+      // 2) Work out how the fee is handled
```

Comment:
do you have a plan for this? I am fearing this changes the setup a lot...

Reply:
Maybe not just the meesage but the tokens should also be send via the bridge... maybe its not too much changing the design

Reply:
So I need to look at how the token bridging works, I'm not sure about that part.

I'm also not totally sure if there's a generic way to check later that the message receipt reverted, but I guess not. If not  we can check the preconditions for the fork such that the call to `requestFork` will never revert (if you give it enough gas) and instead store the result. If the result is that the preconditions weren't handled, we can bridge that information back and call `cancelArbitration` which is a feature I previously added for Kleros when dealing with cross-chain stuff.

The hitch is that the current ForkingManager code may revert for deployment of both forks if someone tries to deploy one of them as hostile. We already need to change ForkingManager so that the requesting just sends the fee to the auction and starts the countdown on the fork, and the actual activation of the fork happens 1 week (or whatever) later. I had a look at solidity exception handling but I can't see a way around making the initialization of the two forks be callable separately, like I did in the previous version (ForkManager) instead of initializing them both in the same call. If we want to we can still create the contracts at the `requestFork` stage so we can see their addresses, but we can't call any initialization code inside that function because it might revert the whole thing.